doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,1452 @@
+it is recommended to use a forward reference to solve this for now.
+
+At import time, the default behavior in Python is to read in all type hints and store them in `__annotations__` as their actual types.
+`from __future__ import annotations` changes this such that type hints don't get evaluated at runtime and are preserved as string literals in the `__annotations__` dictionary.
+There is no difference in behavior for the type checkers with using this import. Note that `from __future__ import annotations` must be imported at the top of the file before any other imports.
+
+It's also worth calling out that using this import also allows use of generic collection type hints like `dict` and `list` instead of `typing.Dict` and `typing.List`. 
+
+More details about this import and its behavior can be found in [PEP 563](https://peps.python.org/pep-0563/). Information
+about the PEP which may supersede this can be found in [PEP 649](https://peps.python.org/pep-0649/).
+
+### Use type aliases to create readable names for types
+
+Use a type alias to create more readable types or convey intent.
+
+```python
+from typing import List, Tuple, Dict
+
+Coordinate = Tuple[float, float]
+Route = List[Coordinate]
+TrailName = str
+Trails = Dict[TrailName, Route]
+
+
+def get_trails(state) -> Trails:
+    ...
+
+# vs.
+
+def get_trails(state) -> Dict[str, List[Tuple[float, float]]]:
+    ...
+```
+
+If it's possible that a type alias could be confused with a global assignment, e.g. `x = 1`, you can use `typing.TypeAlias`.
+This removes ambiguity for the type checker and indicates this isn't a normal variable assignment.
+
+```python
+# from typing import TypeAlias Python >=3.10
+from typing_extensions import TypeAlias
+
+x: TypeAlias = 1
+```
+
+> Note that TypeAlias is backported to older versions of Python by using typing_extensions.
+
+### Use typing.overload to overload a function
+
+`typing.overload` allows for annotating different combinations of typed arguments for a function. This can be useful when the
+return type might depend on the type or types of parameters.
+
+Consider a function which takes in different input types which then inform the output type:
+
+```python
+from typing import Union
+
+
+def analyze_text(text: str, analysis_kind: Union[SentimentAnalysis, EntityRecognition, LanguageDetection]) -> Union[
+    SentimentResult, EntityRecognitionResult, LanguageDetectionResult]:
+    return _analyze(text, analysis_kind)
+```
+
+We can annotate this function with overloads to help inform the type checker which input maps to
+which return type.
+
+```python
+from typing import overload, Union
+
+
+@overload
+def analyze_text(text: str, analysis_kind: LanguageDetection) -> LanguageDetectionResult:
+    ...
+
+
+@overload
+def analyze_text(text: str, analysis_kind: EntityRecognition) -> EntityRecognitionResult:
+    ...
+
+
+@overload
```

Comment:
super extra tiny nit, but based on GVM's rec, should there be one blank line b/w these methods?